File: Sources/MonitorPresentation/MonitorAssistPaletteReveal.swift
import Foundation

/// Interactive View Assist plate: tap the arrow, or press and drag it so the
/// expanding edge stays under the finger. A flick coasts to open or close.
/// 0 is collapsed, 1 is the full catalog.
public enum MonitorAssistPaletteReveal: Sendable {
    public static let snap = 0.5
    public static let slop = 8.0
    /// Flick along the expand axis, in points per second.
    public static let flick = 280.0
    /// Ballistic coast used to project a flick, in seconds.
    public static let coast = 0.22

    public static func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double {
        let clamped = min(1, max(0, t))
        return a + (b - a) * clamped
    }

    /// Portrait expands upward (−Y); landscape expands to the trailing edge (+X).
    public static func translationAlongExpand(dx: Double, dy: Double, portrait: Bool) -> Double {
        portrait ? -dy : dx
    }

    public static func progress(visible: Double, compact: Double, full: Double) -> Double {
        let span = full - compact
        guard span > 1 else { return full >= compact ? 1 : 0 }
        return min(1, max(0, (visible - compact) / span))
    }

    public static func progress(
        translationAlongExpand: Double, span: Double, fromExpanded: Bool
    ) -> Double {
        guard span > 1 else { return fromExpanded ? 1 : 0 }
        let origin = fromExpanded ? span : 0
        return min(1, max(0, (origin + translationAlongExpand) / span))
    }

    /// Keep the grabbed point on the expanding edge under the finger.
    /// Landscape: trailing edge in a leading-origin space (`finger + grabOffset`).
    public static func visibleEdge(finger: Double, grabOffset: Double, compact: Double, full: Double)
        -> Double
    {
        min(full, max(compact, finger + grabOffset))
    }

    /// Portrait expanding edge is the top, at `full - visible` in a top-origin space.
    public static func portraitGrabOffset(
        fingerY: Double, visibleHeight: Double, fullHeight: Double
    ) -> Double {
        fingerY - (fullHeight - visibleHeight)
    }

    public static func portraitVisibleHeight(
        fingerY: Double, grabOffset: Double, compact: Double, full: Double
    ) -> Double {
        min(full, max(compact, full - (fingerY - grabOffset)))
    }

    public static func projectedProgress(
        progress: Double, velocityAlongExpand: Double, span: Double
    ) -> Double {
        guard span > 1 else { return progress }
        return min(1, max(0, progress + velocityAlongExpand * coast / span))
    }

    public static func shouldOpen(
        progress: Double, velocityAlongExpand: Double, projectedProgress: Double? = nil
    ) -> Bool {
        if abs(velocityAlongExpand) >= flick { return velocityAlongExpand > 0 }
        return (projectedProgress ?? progress) >= snap
    }

    /// Tools beyond the collapsed favorites fade their glyphs only.
    public static func extraToolOpacity(progress: Double) -> Double {
        min(1, max(0, (progress - 0.08) / 0.42))
    }

    public static func landscapeCellIndex(column: Int, row: Int) -> Int {
        column * 2 + row
    }

    public static func compactToolCount(portrait: Bool) -> Int {
        portrait ? 1 : 2
    }

    /// Keep the open plate still. Ranking updates apply only after collapse.
    public static func pinnedOrder(ranked: [String], pinned: [String], frozen: Bool) -> [String] {
        guard frozen, !pinned.isEmpty else { return ranked }
        var seen: Set<String> = []
        let kept = pinned.filter { ranked.contains($0) && seen.insert($0).inserted }
        let extras = ranked.filter { seen.insert($0).inserted }
        return kept + extras
    }
}

File: Apps/Android/monitor-ui/src/main/kotlin/com/opencapture/monitorui/MonitorAssistPaletteReveal.kt
package com.opencapture.monitorui

import androidx.compose.ui.unit.IntRect
import androidx.compose.ui.unit.IntSize
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/** Interactive View Assist plate: tap the arrow, or press and drag it. */
object MonitorAssistPaletteReveal {
    const val SNAP = 0.5f
    const val SLOP = 8f
    const val FLICK = 280f
    const val COAST = 0.22f

    /** Reserve the full plate before its first reveal frame and through the collapse spring. */
    fun reservationSize(expanded: Boolean, dragging: Boolean, animating: Boolean, progress: Float,
        compact: IntSize, full: IntSize): IntSize =
        if (expanded || dragging || animating || progress > 0.001f) full else compact

    /** The same bottom-leading window placement drives rendering and obstruction bounds. */
    fun popupBounds(slot: IntRect, windowWidth: Int, size: IntSize): IntRect {
        val left = slot.left.coerceIn(0, maxOf(0, windowWidth - size.width))
        val top = (slot.bottom - size.height).coerceAtLeast(0)
        return IntRect(left, top, left + size.width, top + size.height)
    }

    fun lerp(a: Float, b: Float, t: Float): Float {
        val clamped = min(1f, max(0f, t))
        return a + (b - a) * clamped
    }

    fun translationAlongExpand(dx: Float, dy: Float, portrait: Boolean): Float =
        if (portrait) -dy else dx

    fun progress(visible: Float, compact: Float, full: Float): Float {
        val span = full - compact
        if (span <= 1f) return if (full >= compact) 1f else 0f
        return min(1f, max(0f, (visible - compact) / span))
    }

    fun progress(translationAlongExpand: Float, span: Float, fromExpanded: Boolean): Float {
        if (span <= 1f) return if (fromExpanded) 1f else 0f
        val origin = if (fromExpanded) span else 0f
        return min(1f, max(0f, (origin + translationAlongExpand) / span))
    }

    fun visibleEdge(finger: Float, grabOffset: Float, compact: Float, full: Float): Float =
        min(full, max(compact, finger + grabOffset))

    fun portraitGrabOffset(fingerY: Float, visibleHeight: Float, fullHeight: Float): Float =
        fingerY - (fullHeight - visibleHeight)

    fun portraitVisibleHeight(fingerY: Float, grabOffset: Float, compact: Float, full: Float): Float =
        min(full, max(compact, full - (fingerY - grabOffset)))

    /** Top-origin Y in the full plate, matching iOS `assistPalette` space. */
    fun paletteFingerY(fingerScreenY: Float, plateBottomScreenY: Float, fullHeight: Float): Float =
        fingerScreenY - (plateBottomScreenY - fullHeight)

    fun projectedProgress(progress: Float, velocityAlongExpand: Float, span: Float): Float {
        if (span <= 1f) return progress
        return min(1f, max(0f, progress + velocityAlongExpand * COAST / span))
    }

    fun shouldOpen(progress: Float, velocityAlongExpand: Float, projectedProgress: Float = progress): Boolean {
        if (abs(velocityAlongExpand) >= FLICK) return velocityAlongExpand > 0f
        return projectedProgress >= SNAP
    }

    fun extraToolOpacity(progress: Float): Float = min(1f, max(0f, (progress - 0.08f) / 0.42f))

    fun landscapeCellIndex(column: Int, row: Int): Int = column * 2 + row

    fun compactToolCount(portrait: Boolean): Int = if (portrait) 1 else 2

    /** Keep the open plate still. Ranking updates apply only after collapse. */
    fun pinnedOrder(ranked: List<String>, pinned: List<String>, frozen: Boolean): List<String> {
        if (!frozen || pinned.isEmpty()) return ranked
        val seen = linkedSetOf<String>()
        val kept = pinned.filter { it in ranked && seen.add(it) }
        val extras = ranked.filter { seen.add(it) }
        return kept + extras
    }
}
